code/game/objects/structures/lavaland/ore_vent.dm

```diff
@@ -0,0 +1,526 @@
+		return
+	if(!COOLDOWN_FINISHED(src, wave_cooldown))
+		balloon_alert_to_viewers("protect the node drone!")
+		return
+	if(scan_only)
+		discovered = TRUE
+		generate_description(user)
+		balloon_alert_to_viewers("vent scanned!")
+		return
+	if(!discovered)
+		balloon_alert(user, "scanning...")
+		playsound(src, 'sound/items/timer.ogg', 30, TRUE)
+		if(do_after(user, 4 SECONDS))
+			discovered = TRUE
+			balloon_alert(user, "vent scanned!")
+		generate_description(user)
+		var/obj/item/card/id/user_id_card = user.get_idcard(TRUE)
+		if(isnull(user_id_card))
+			return
+		user_id_card.registered_account.mining_points += (MINER_POINT_MULTIPLIER)
+		user_id_card.registered_account.bank_card_talk("You've been awarded [MINER_POINT_MULTIPLIER] mining points for discovery of an ore vent.")
+		return
+
+	if(tgui_alert(user, excavation_warning, "Begin defending ore vent?", list("Yes", "No")) != "Yes")
+		return
+	if(!COOLDOWN_FINISHED(src, wave_cooldown))
+		return
+	//This is where we start spitting out mobs.
+	Shake(duration = 3 SECONDS)
+	node = new /mob/living/basic/node_drone(loc)
+	node.arrive(src)
+	RegisterSignal(node, COMSIG_QDELETING, PROC_REF(handle_wave_conclusion))
+	particles = new /particles/smoke/ash()
+
+	for(var/i in 1 to 5) // Clears the surroundings of the ore vent before starting wave defense.
+		for(var/turf/closed/mineral/rock in oview(i))
+			if(istype(rock, /turf/open/misc/asteroid) && prob(35)) // so it's too common
+				new /obj/effect/decal/cleanable/rubble(rock)
+			if(prob(100 - (i * 15)))
+				rock.gets_drilled(user, FALSE)
+				if(prob(50))
+					new /obj/effect/decal/cleanable/rubble(rock)
+		sleep(0.6 SECONDS)
+
+	start_wave_defense()
+
+/**
+ * Generates a description of the ore vent to ore_string, based on the minerals contained within it.
+ * Ore_string is passed to examine().
+ */
+/obj/structure/ore_vent/proc/generate_description(mob/user)
+	for(var/mineral_count in 1 to mineral_breakdown.len)
+		var/datum/material/resource = mineral_breakdown[mineral_count]
+		if(mineral_count == mineral_breakdown.len)
+			ore_string += "and " + span_bold(initial(resource.name)) + "."
+		else
+			ore_string += span_bold(initial(resource.name)) + ", "
+	if(user)
+		ore_string += "\nThis vent was first discovered by [user]."
+/**
+ * Adds floating temp_visual overlays to the vent, showcasing what minerals are contained within it.
+ * If undiscovered, adds a single overlay with the icon_state "unknown".
+ */
+/obj/structure/ore_vent/proc/add_mineral_overlays()
+	if(mineral_breakdown.len && !discovered)
+		var/obj/effect/temp_visual/mining_overlay/vent/new_mat = new /obj/effect/temp_visual/mining_overlay/vent(drop_location())
+		new_mat.icon_state = "unknown"
+		return
+	for(var/datum/material/selected_mat as anything in mineral_breakdown)
+		var/obj/effect/temp_visual/mining_overlay/vent/new_mat = new /obj/effect/temp_visual/mining_overlay/vent(drop_location())
+		new_mat.icon_state = selected_mat.name
```

Comment:
Runtiming: Cannot read null.name, related to mineral scan pulse such as from a mining scanner.